### PR TITLE
experimental: Benchmark reconciling tables and BPF

### PR DESCRIPTION
**Pull request by DamianSawicki**
Add a benchmark for addition of k8s services and endpoints, reflecting them to StateDB tables, and reconciliation to BPF maps.

 ---

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

### CODE REVIEW

**Comment by robscott** [comment]
cc @bowei

File: pkg/loadbalancer/experimental/k8s_benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/testutils"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+func BenchmarkServiceAndEndpoint(t *testing.B) {
+	// As we're using k8s.Endpoints we need to set this to ask ParseEndpoint*
```
Comment (bowei) [inline pkg/loadbalancer/experimental/k8s_benchmark_test.go]: As this test modifies global variables -- we should put a comment at the top of the test saying this cannot be parallelized:

`// No t.parallel(): modified global variables.`

File: pkg/loadbalancer/experimental/k8s_benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+			)
+			h := createHive(maps, services, pods, endpoints, 0.1, &writer, &db, &bo) // 10% failure probability for lbmaps
+			// TODO(Jussi): without non-zero failure probability the benchmark sometimes fails to clean up BPF when log level is Info.
+
+			require.NoError(t, h.Start(log, context.TODO()))
+			t.Cleanup(func() {
+				assert.NoError(t, h.Stop(log, context.TODO()))
+			})
+
+			memChanges := []*memoryPair{}
```
Comment (bowei) [inline pkg/loadbalancer/experimental/k8s_benchmark_test.go]: var memChanges []*memoryPair

(prefer nil list to empty list)

File: pkg/loadbalancer/experimental/k8s_benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+			for i := 0; i < t.N; i++ {
+				runtime.GC()
+				var memory memoryPair
+				runtime.ReadMemStats(&memory.before)
+
+				if tc.timeInsertion {
+					t.StartTimer()
+				}
+
+				//
```
Comment (bowei) [inline pkg/loadbalancer/experimental/k8s_benchmark_test.go]: [minor] remove extra `// (empty)` comments

File: pkg/loadbalancer/experimental/k8s_benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+type memoryPair struct {
+	before runtime.MemStats
+	after  runtime.MemStats
+}
+
+func printMemoryStats(t *testing.B, pairs []*memoryPair, testSize int) {
+	type stats struct {
+		objects, alloc, inUse int64
+	}
+	var Min, Max, Avg stats
```
Comment (bowei) [inline pkg/loadbalancer/experimental/k8s_benchmark_test.go]: Suggest pulling out the aggregation logic into a separate function.

File: pkg/loadbalancer/experimental/k8s_load_test.go [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/testutils"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+func TestServiceAndEndpoint(t *testing.T) {
+	// As we're using k8s.Endpoints we need to set this to ask ParseEndpoint*
```
Comment (bowei) [inline pkg/loadbalancer/experimental/k8s_load_test.go]: // No t.Parallel(): modifies global variables.

File: pkg/loadbalancer/experimental/k8s_load_test.go [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+	// to handle the termination state. Eventually this should migrate to the
+	// package for the k8s data source.
+	oldEnableK8sTerminatingEndpoint := option.Config.EnableK8sTerminatingEndpoint
+	defer func() {
+		option.Config.EnableK8sTerminatingEndpoint = oldEnableK8sTerminatingEndpoint
+	}()
+	option.Config.EnableK8sTerminatingEndpoint = true
+
+	const testSize = 10000
```
Comment (bowei) [inline pkg/loadbalancer/experimental/k8s_load_test.go]: add // comment on what testSize represents

**Review by bowei — COMMENTED** [review]
some minor comments.

You may want to consider having two commits -- one move and another that does diffs.

File: pkg/loadbalancer/experimental/k8s_benchmark_test.go (hunk 1 above)
Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/k8s_benchmark_test.go]: I added it in the other file, but this one is a benchmark, and `*testing.B` has no method `Parallel()`. I renamed `t -> b` to make it more explicit that this is a benchmark.

File: pkg/loadbalancer/experimental/k8s_benchmark_test.go (hunk 2 above)
Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/k8s_benchmark_test.go]: Done.

File: pkg/loadbalancer/experimental/k8s_benchmark_test.go (hunk 3 above)
Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/k8s_benchmark_test.go]: This is taken verbatim from `pkg/loadbalancer/experimental/k8s_test.go`.

File: pkg/loadbalancer/experimental/k8s_benchmark_test.go (hunk 4 above)
Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/k8s_benchmark_test.go]: Done.

File: pkg/loadbalancer/experimental/k8s_load_test.go (hunk 5 above)
Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/k8s_load_test.go]: Done.

File: pkg/loadbalancer/experimental/k8s_load_test.go (hunk 6 above)
Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/k8s_load_test.go]: Added.

**Comment by DamianSawicki** [comment]
> some minor comments.
> 
> You may want to consider having two commits -- one move and another that does diffs.

Thanks for review @bowei! I've split this into two commits as you suggested.

**Review by joamaki — APPROVED** [review]

**Pull request merged**